subprojects/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/processor/ConfigurationCacheInstrumentationProcessorTest.groovy

```diff
@@ -0,0 +1,163 @@
+
+    def "code generation is deterministic regardless of type processing order"() {
+        given:
+        withAnnotatedType 'Foo'
+        withAnnotatedType 'Bar'
+        def annotatedTypes = ['my.Foo', 'my.Bar']
+
+        when:
+        DiagnosticCollector<JavaFileObject> diagnostics = processAnnotationsOf(annotatedTypes)
+
+        then:
+        errorsFrom(diagnostics).isEmpty()
+
+        and:
+        def outputFile = outDir.file('my/Interceptors.java')
+        outputFile.file
+
+        when:
+        def originalOutput = outputFile.text
+
+        and:
+        processAnnotationsOf(annotatedTypes.reverse())
+
+        then:
+        outputFile.text == originalOutput
+    }
+
+    def "code generation is deterministic regardless of method declaration order"() {
+        given:
+        withMethodDeclarationOrder 'foo', 'bar'
+
+        when:
+        DiagnosticCollector<JavaFileObject> diagnostics = processAnnotationsOf(['my.Test'])
+
+        then:
+        errorsFrom(diagnostics).isEmpty()
+
+        and:
+        def outputFile = outDir.file('my/Interceptors.java')
+        outputFile.file
+
+        when:
+        def originalOutput = outputFile.text
+        withMethodDeclarationOrder 'bar', 'foo'
+
+        and:
+        processAnnotationsOf(['my.Test'])
+
+        then:
+        outputFile.text == originalOutput
+    }
+
+    private withAnnotatedType(String typeName) {
+        sourceFile "my/${typeName}.java", """
+            package my;
+
+            import org.gradle.internal.instrumentation.api.annotations.*;
+
+            @SpecificJvmCallInterceptors(generatedClassName = "my.Interceptors")
+            class $typeName {
+                @InterceptCalls
+                @CallableKind.StaticMethod(ofClass = my.${typeName}.class)
+                @CallableDefinition.Name("stubTarget")
+                public static void stub() {}
+            }
+        """
+    }
+
+    private void withMethodDeclarationOrder(String first, String second) {
+        sourceFile 'my/Test.java', """
+            package my;
+
+            import org.gradle.internal.instrumentation.api.annotations.*;
+
+            @SpecificJvmCallInterceptors(generatedClassName = "my.Interceptors")
+            class Test {
+                @InterceptCalls
+                @CallableKind.StaticMethod(ofClass = my.Test.class)
+                @CallableDefinition.Name("${first}Target")
+                public static void ${first}() {}
+
+                @InterceptCalls
+                @CallableKind.StaticMethod(ofClass = my.Test.class)
+                @CallableDefinition.Name("${second}Target")
+                public static void ${second}() {}
+            }
+        """
+    }
+
+    private void sourceFile(String name, @Language('java') String source) {
+        srcDir.file(name).text = source.stripIndent()
+    }
+
+    private DiagnosticCollector<JavaFileObject> processAnnotationsOf(Iterable<String> classes) {
```

Comment:
👍 Wow, I did not expect the setup for the test to be so compact!

Reply:
ChatGPT did it 😉